crates/oidc-client/src/requests/authorization_code.rs

```diff
@@ -0,0 +1,441 @@
+// Copyright 2022 Kévin Commaille.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Requests for the [Authorization Code flow].
+//!
+//! [Authorization Code flow]: https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth
+
+use base64ct::{Base64UrlUnpadded, Encoding};
+use headers::HeaderValue;
+use http::header::CONTENT_TYPE;
+use mas_http::{CatchHttpCodesLayer, FormUrlencodedRequestLayer, JsonResponseLayer};
+use mas_iana::oauth::{OAuthAuthorizationEndpointResponseType, PkceCodeChallengeMethod};
+use mas_jose::claims::{self, TokenHash};
+use oauth2_types::{
+    pkce,
+    prelude::CodeChallengeMethodExt,
+    requests::{
+        AccessTokenRequest, AccessTokenResponse, AuthorizationCodeGrant, AuthorizationRequest,
+        Prompt, PushedAuthorizationResponse,
+    },
+    scope::Scope,
+};
+use rand::{
+    distributions::{Alphanumeric, DistString},
+    thread_rng, Rng,
+};
+use serde::Serialize;
+use serde_with::skip_serializing_none;
+use tower::{Layer, Service, ServiceExt};
+use url::Url;
+
+use super::jose::JwtVerificationData;
+use crate::{
+    error::{
+        AuthorizationError, IdTokenError, PushedAuthorizationError, TokenAuthorizationCodeError,
+    },
+    http_service::HttpService,
+    requests::{jose::verify_id_token, token::request_access_token},
+    types::{
+        client_credentials::ClientCredentials,
+        scope::{ScopeExt, ScopeToken},
+        IdToken,
+    },
+    utils::{http_all_error_status_codes, http_error_mapper},
+};
+
+/// The data necessary to build an authorization request.
+pub struct AuthorizationRequestData<'a> {
+    /// The ID obtained when registering the client.
+    pub client_id: String,
+
+    /// The PKCE methods supported by the issuer, from its metadata.
+    pub code_challenge_methods_supported: &'a Option<Vec<PkceCodeChallengeMethod>>,
```

Comment:
Well I made everything owned when I needed an owned type and this is a ref because I only need a ref :sweat_smile:

Reply:
We could have a lot more borrowed types in general if the oauth2-types `Request`s accepted borrowed types, because we don't need owned types for serializing. That would however mean we need two types for each, one borrowed (for sending) and one owned (for receiving).